From bbe2675f18d99837e400c7a1bbb28241d86cbe15 Mon Sep 17 00:00:00 2001
From: Murali Karicheri <m-karicheri2@ti.com>
Date: Mon, 30 Oct 2017 15:27:47 -0400
Subject: [PATCH 95/97] hsr: fix incorrect lsdu size in the tag of HSR frames
 for small frames

For small Ethernet frames with size less than minimum size
66 for HSR/PRP vs 60 for regular Ethernet frames, hsr driver currently
doesn't pad the frame to make it minimum size. This results in incorrect
LSDU size being populated in the HSR tag for these frames. Fix this by
padding the frame to the minimum size applicable for HSR.

Signed-off-by: Murali Karicheri <m-karicheri2@ti.com>
---
 net/hsr-prp/hsr_prp_forward.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/hsr-prp/hsr_prp_forward.c b/net/hsr-prp/hsr_prp_forward.c
index bbc2642..736e6d4 100644
--- a/net/hsr-prp/hsr_prp_forward.c
+++ b/net/hsr-prp/hsr_prp_forward.c
@@ -234,6 +234,9 @@ static void hsr_fill_tag(struct sk_buff *skb, struct hsr_prp_frame_info *frame,
 	unsigned char *pc;
 	int lsdu_size;
 
+	/* pad to minimum packet size which is 60 + 6 (HSR tag) */
+	skb_put_padto(skb, ETH_ZLEN + HSR_PRP_HLEN);
+
 	lsdu_size = skb->len - 14;
 	if (frame->is_vlan)
 		lsdu_size -= 4;
-- 
1.9.1

